Library/OcCryptoLib/RsaDigitalSign.c

```diff
@@ -0,0 +1,511 @@
+  found in the LICENSE file.
+
+  Implementation of RSA signature verification which uses a pre-processed key
+  for computation.
+**/
+
+#ifdef EFIAPI
+#include <Library/BaseMemoryLib.h>
+#endif
+
+#include <Library/DebugLib.h>
+#include <Library/BaseLib.h>
+#include <Library/OcGuardLib.h>
+#include <Library/OcCryptoLib.h>
+
+/**
+  PKCS#1 padding (from the RSA PKCS#1 v2.1 standard)
+
+  The DER-encoded padding is defined as follows :
+  0x00 || 0x01 || PS || 0x00 || T
+
+  T: DER Encoded DigestInfo value which depends on the hash function used,
+  for SHA-256:
+  (0x)30 31 30 0d 06 09 60 86 48 01 65 03 04 02 01 05 00 04 20 || H.
+
+  Length(T) = 51 octets for SHA-256
+
+  PS: octet string consisting of {Length(RSA Key) - Length(T) - 3} 0xFF
+ **/
+#define PKCS_PAD_SIZE (CONFIG_RSA_KEY_SIZE - SHA256_DIGEST_SIZE)
+
+STATIC  UINT8 mSha256Tail[] = {
+  0x00, 0x30, 0x31, 0x30, 0x0d, 0x06, 0x09, 0x60,
+  0x86, 0x48, 0x01, 0x65, 0x03, 0x04, 0x02, 0x01,
+  0x05, 0x00, 0x04, 0x20
+};
+
+UINT64
+Mula32 (
+  UINT32  A,
+  UINT32  B,
+  UINT32  C
+  )
+{
+  UINT64 Ret = A;
+
+  Ret *= B;
+  Ret += C;
+  return Ret;
+}
+
+UINT64
+Mulaa32 (
+  UINT32  A,
+  UINT32  B,
+  UINT32  C,
+  UINT32  D
+  )
+{
+  UINT64 Ret = A;
+
+  Ret *= B;
+  Ret += C;
+  Ret += D;
+  return Ret;
+}
+
+//
+//  A[] -= Mod
+//
+STATIC
+VOID
+SubMod (
+  RSA_PUBLIC_KEY  *Key,
+  UINT32          *A
+  )
+{
+  INT64  B     = 0;
+  UINT32 Index = 0;
+  for (Index = 0; Index < Key->Size; Index++) {
+    B += (UINT64) A[Index] - Key->N[Index];
+    A[Index] = (UINT32) B;
+    B >>= 32;
+  }
+}
+
+//
+// Return A[] >= Mod
+//
+STATIC
+INT32
+GeMod (
+  RSA_PUBLIC_KEY  *Key,
+  CONST UINT32    *A
+  )
+{
+  UINT32 Index = 0;
+
+  for (Index = Key->Size; Index;) {
+    Index--;
+    if (A[Index] < Key->N[Index]){
+      return 0;
+    }
+    if (A[Index] > Key->N[Index]){
+      return 1;
+    }
+  }
+  return 1;
+}
+
+//
+// Montgomery c[] += a * b[] / R % mod
+//
+STATIC
+VOID
+MontMulAdd (
+  RSA_PUBLIC_KEY  *Key,
+  UINT32          *C,
+  UINT32          Aa,
+  UINT32          *Bb
+  )
+{
+  UINT64 A, B;
+  UINT32 D0, Index;
+
+  A = Mula32 (Aa, Bb[0], C[0]);
+  D0 = (UINT32) A * Key->N0Inv;
+  B = Mula32 (D0, Key->N[0], (UINT32) A);
+
+  for (Index = 1; Index < Key->Size; Index++) {
+    A = Mulaa32 (Aa, Bb[Index], C[Index], (UINT32) (A >> 32));
+    B = Mulaa32 (D0, Key->N[Index], (UINT32) A, (UINT32) (B >> 32));
+    C[Index - 1] = (UINT32) B;
+  }
+
+  A = (A >> 32) + (B >> 32);
+  C[Index - 1] = (UINT32) A;
+
+  if (A >> 32) {
+    SubMod (Key, C);
+  }
+}
+
+//
+// Montgomery c[] = a[] * b[] / R % mod
+//
+STATIC
+VOID
+MontMul (
+  RSA_PUBLIC_KEY  *Key,
+  UINT32          *C,
+  UINT32          *A,
+  UINT32          *B
+  )
+{
+  UINT32 Index;
+
+  ZeroMem (C, Key->Size);
+
+  for (Index = 0; Index < Key->Size; Index++) {
+    MontMulAdd (Key, C, A[Index], B);
+  }
+}
+
+/**
+  In-place public exponentiation.
+  Exponent depends on the configuration (65537 (default), or 3).
```

Comment:
The comment mentions either 65537 or 3 exponent, however I only see code for 65537.